### DISCUSSION TITLE
Setting up DynamoDB with Laravel

### DISCUSSION THREAD

**Discussion by architechpro** [brefphp/bref#1195] [Support] [answered]
Hi,

I've been setting up Laravel 9 on bref and having a problem with making any basic operation with cache (like `$value = Cache::get('value');`) after finishing the setup:

```
Aws\DynamoDb\Exception\DynamoDbException
Error executing "GetItem" on "dynamodb.us-east-2.amazonaws.com"; AWS HTTP error: Client error: `POST dynamodb.us-east-2.amazonaws.com` resulted in a `400 Bad Request` response:
{"__type":"com.amazon.coral.service#InvalidSignatureException","message":"'Host' or ':authority' must be a 'SignedHeader (truncated...)
 InvalidSignatureException (client): 'Host' or ':authority' must be a 'SignedHeader' in the AWS Authorization. - {"__type":"com.amazon.coral.service#InvalidSignatureException","message":"'Host' or ':authority' must be a 'SignedHeader' in the AWS Authorization."}
```

I double-checked that:
- aws/aws-sdk-php is installed;
- the config/cache.php is correct and contains all required values, dynamodb is set as default cache driver;
- app/Http/Middleware/TrustProxies.php has `protected $proxies = '*';` as recommended by [bref Laravel docs](https://bref.sh/docs/frameworks/laravel.html#trusted-proxies);
- dynamodb part of the config is set according to bref documentation ([general](https://bref.sh/docs/environment/storage.html#deploying-dynamodb-tables) and [laravel-specific](https://bref.sh/docs/frameworks/laravel.html#using-dynamodb)):

```
    'dynamodb' => [
            'driver' => 'dynamodb',
            'key' => env('AWS_ACCESS_KEY_ID'),
            'secret' => env('AWS_SECRET_ACCESS_KEY'),
            'region' => env('AWS_DEFAULT_REGION', 'us-east-1'),
            'table' => env('DYNAMODB_CACHE_TABLE', 'cache'),
            'endpoint' => env('DYNAMODB_ENDPOINT'),
            'attributes' => [
                'key' => 'id',
                'expiration' => 'ttl',
            ]  
        ],
```

As far as I know, headers should be automatically set by Aws classes under the hood. I can see that lots of people are successfully using DynamoDB with Bref so I'll appreciate any suggestions on setting everything up correctly. Thanks!

**UPD**: I've created a simple test function that uses DynamoDbClient class directly and it works well with API keys I provide in manually created env vars. 

```
   use Aws\DynamoDb\DynamoDbClient;
   
   ...

    public function test()
    {
        $client = new DynamoDbClient([
            'region'  => 'us-east-2',
            'version' => 'latest',
            'credentials' => [
              'key'    => env("AWS_BREF_ACCESS_KEY"),
              'secret' => env("AWS_BREF_SECRET_KEY"),
            ]
        ]);

        $result = $client->describeTable(array(
            'TableName' => env("DYNAMODB_CACHE_TABLE")
        ));

        echo '<pre>'; var_dump($result); echo '</pre>';
        exit();
    }
```

So it seems that the issue is with using DynamoDB through Laravel Cache facade.

**Comment by mnapoli** [comment]
Hi, it may be the same problem as with the S3 configuration:

<img width="739" alt="image" src="https://user-images.githubusercontent.com/720328/165084505-31fd057b-db5d-4507-bb29-57ba8c086474.png">

(from the docs: https://bref.sh/docs/frameworks/laravel.html#file-storage-on-s3)

I would add this line in the DynamoDB config:

```
           'token' => env('AWS_SESSION_TOKEN'),
```

https://github.com/laravel/laravel/pull/5138#issuecomment-660691980

**Reply by architechpro** [reply]
Thank you for the suggestion! Unfortunately, that didn't change anything.

I have created a clean project from scratch and didn't make any changes other than DynamoDB config. And DynamoDB worked. My guess about possible reasons:
- I tried to set up Redis before DynamoDB, maybe some kind of a conflict is caused by previously set up and then removed Redis packages;
- I have set up a new clean test project in a default region, us-east-1. I guess that the fact that I used a non-default region for a setup caused a problem.

So I'm resolving the problem by moving all the code to a new clean project in a default AWS region.

**Comment by pmayet** [comment]
Hi,

You can check my config :
https://github.com/pmayet/laravel-lambda-5min/tree/dynamodb-session-improve

An article :
***@***.***/d%C3%A9ployer-une-application-laravel-serverless-avec-stockage-des-sessions-et-du-cache-applicatif-dans-2aed4ac8c815

Best

Patrice

Le mar. 26 avr. 2022 à 12:35, architechpro ***@***.***> a
écrit :

**Comment by architechpro** [comment, marked as answer]
I have created a clean project from scratch and didn't make any changes other than DynamoDB config. And DynamoDB worked. My guess about possible reasons:

I tried to set up Redis before DynamoDB, maybe some kind of a conflict is caused by previously set up and then removed Redis packages;
I have set up a new clean test project in a default region, us-east-1. I guess that the fact that I used a non-default region for a setup caused a problem.
So I'm resolving the problem by moving all the code to a new clean project in a default AWS region.
